metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java

```diff
@@ -0,0 +1,262 @@
+    private final FaultHandler faultHandler;
+
+    /**
+     * The minimum number of bytes we will wait to see before emitting a snapshot.
+     */
+    private final long minBytesSinceLastSnapshot;
+
+    /**
+     * The minimum amount of time we will wait before emitting a snapshot, or 0 to disable
+     * time-based snapshotting.
+     */
+    private final long minTimeSinceLastSnapshotNs;
+
+    /**
+     * If non-null, the reason why snapshots have been disabled.
+     */
+    private volatile String disabledReason;
+
+    /**
+     * The event queue used to schedule emitting snapshots.
+     */
+    private final EventQueue eventQueue;
+
+    /**
+     * The log bytes that we have read since the last snapshot.
+     */
+    private long bytesSinceLastSnapshot;
+
+    /**
+     * The time at which we created the last snapshot.
+     */
+    private long lastSnapshotTimeNs;
+
+    private SnapshotGenerator(
+        int nodeId,
+        Time time,
+        Emitter emitter,
+        FaultHandler faultHandler,
+        long minBytesSinceLastSnapshot,
+        long minTimeSinceLastSnapshotNs
+    ) {
+        this.nodeId = nodeId;
+        this.time = time;
+        this.emitter = emitter;
+        this.faultHandler = faultHandler;
+        this.minBytesSinceLastSnapshot = minBytesSinceLastSnapshot;
+        this.minTimeSinceLastSnapshotNs = minTimeSinceLastSnapshotNs;
+        LogContext logContext = new LogContext("[SnapshotGenerator " + nodeId + "] ");
+        this.log = logContext.logger(SnapshotGenerator.class);
+        this.disabledReason = null;
+        this.eventQueue = new KafkaEventQueue(time, logContext, "");
+        resetSnapshotCounters();
+    }
+
+    @Override
+    public String name() {
+        return "SnapshotGenerator";
+    }
+
+    void resetSnapshotCounters() {
+        this.bytesSinceLastSnapshot = 0L;
+        this.lastSnapshotTimeNs = time.nanoseconds();
+    }
+
+    @Override
+    public void publishSnapshot(
+        MetadataDelta delta,
+        MetadataImage newImage,
+        SnapshotManifest manifest
+    ) {
+        // Reset the snapshot counters because we just read a snapshot.
+        resetSnapshotCounters();
+    }
+
+    @Override
+    public void publishPreVersionChangeImage(
+        MetadataDelta delta,
+        MetadataImage preVersionChangeImage,
+        PreVersionChangeManifest manifest
+    ) {
+        scheduleEmit("the metadata version changed", preVersionChangeImage);
+    }
+
+    @Override
+    public void publishLogDelta(
+        MetadataDelta delta,
+        MetadataImage newImage,
+        LogDeltaManifest manifest
+    ) {
+        bytesSinceLastSnapshot += manifest.numBytes();
+        if (bytesSinceLastSnapshot >= minBytesSinceLastSnapshot) {
+            if (eventQueue.isEmpty()) {
+                scheduleEmit("we have replayed at least " + minBytesSinceLastSnapshot +
+                    " bytes", newImage);
+            }
+        } else if (minTimeSinceLastSnapshotNs != 0 &&
```

Comment:
I thought the intent was to disable time-based snapshots with -1.

Reply:
Hmm [KIP-876](https://cwiki.apache.org/confluence/display/KAFKA/KIP-876%3A+Time+based+cluster+metadata+snapshots) currently says 0 means disabled ...